Draft-Accepted/RFCXXXX-Using-IDisposables.md

```diff
@@ -0,0 +1,399 @@
+The proposed syntax for supporting this scenario would be:
+
+```powershell
+using (
+    $fs = [System.IO.File]::OpenRead($Path)
+    $cryptoStream = [System.Security.CryptoGraphy.CryptoStream]::new(
+        $fs,
+        [System.Security.Cryptography.ToBase64Transform]::new(),
+        [System.Security.Cryptography.CryptoStreamMode]::Read)
+    $sr = [System.IO.StreamReader]::new($cryptoStream, [System.Text.Encoding]::ASCII)
+) {
+    $sr.ReadToEnd()
+}
+```
+
+In this scenario the `using ()` statement supports multiple `AssignmentStatementAst` statements allowing it to track multiple variables to dispose.
+The equivalent syntax in C# would be this example:
+
+```csharp
+using (FileStream fs = File.OpenRead(path))
+using (CryptoStream cryptoStream = new CryptoStream(...))
+using (StreamReader sr = new StreamReader(cryptoStream, Encoding.ASCII))
+{
+    sr.ReadToEnd();
+}
+```
+
+## Specification
+
+The syntax for the proposed `Using` statement would be:
+
+```
+using ($<disposable>; ...) {
+    <Statement list>
+}
+```
+
+The part inside the parenthesis is a variable to dispose after the `<Statement list>` ends.
+It can contain multiple variables as well as the code used to define the variable.
+Some examples of this would be:
+
+```powershell
+# Creates the instance separately from the using statement
+$key = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+using ($key) {
+    $key.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+}
+
+# Creates the variable inside the parenthesis
+using ($key = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment') {
+    $key.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+}
+
+# Disposes multiple values at the end of the block
+$machineKey = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+$userKey = Get-Item 'HKCU:\Environment'
+
+using ($machineKey; $userKey) {
+    $m = $machineKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    $u = $userKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    "${m};${u}"
+}
+
+# Same as the above but with the vars defined in the parenthesis
+using (
+    $machineKey = Get-Item 'HKLM:\System\CurrentControlSet\Control\Session Manager\Environment'
+    $userKey = Get-Item 'HKCU:\Environment'
+) {
+    $m = $machineKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    $u = $userKey.GetValue('PATH', '', 'DoNotExpandEnvironmentNames')
+    "${m};${u}"
+}
+```
+
+## Implementation Details
+
+I propose adding a new AST type `UsingDisposableStatementAst` that contains
+
++ `PipelineBaseAst[]` - Disposables
++ `StatementBlockAst` - Body
+
+When visited the Ast will visit the `Disposables` to find any `AssignmentStatementAst` or `VariableExpressionAst` which it can use to keep track of variables to dispose.
+The compiler will then emit an Expression that wraps the `Body` with the below for each `Disposables` (in definition order):
```

Comment:
I don't see how this would affect it as `Invoke-Command` is a blocking call and the session isn't a disposable object anyway. Even if it was I picture your above example to be the equivalent to the following pwsh code

```powershell
try {
    $s = new-pssession
    invoke-command -session $s { <statement list> }
}
finally {
    if ($s) { $s.Dispose() }
}
```